### PR TITLE
Convert test equality to sets to allow out-of-order execution

### PR DESCRIPTION
**Pull request by gdevenyi**
Convert tests to use sets.

Will manually re-run travis a few times to see if this fixes the random errors.

### CODE REVIEW

**Comment by gdevenyi** [comment]
Hrm, I'm not getting the failure mode here. By eye the outputs look the same. Plus now I'm using set equality, though I suspect the set isn't getting constructed correctly by breaking on \n

File: test/test_qbatch.py [hunk 1]
```diff
@@ -64,7 +64,7 @@ def test_run_qbatch_sge_dryrun_array_piped_chunks():
 
         assert array_pipe.returncode == 0, \
             "Chunk {0}: return code = {1}".format(chunk, array_pipe.returncode)
-        assert out.decode('UTF-8') == expected, \
+        assert set(out.decode('UTF-8')) == set(expected), \
```
Comment (pipitone) [inline test/test_qbatch.py]: You'll need to add `.split('\n')` here, no?

**Comment by gdevenyi** [comment]
Blah, splitting didn't fix it :(

Still works locally as well

**Comment by pipitone** [comment]
For whatever reason, one command output is being dropped, e.g:
```
AssertionError: Chunk 3: Expected echo 20	20

echo 21	21

echo 22	22

echo 23	23

echo 24	24

echo 25	25

echo 26	26

echo 27	27 <<---- this guy is missing

echo 28	28

echo 29	29

 but got echo 20	20

echo 21	21

echo 22	22

echo 23	23

echo 24	24

echo 25	25

echo 26	26

echo 28	28

echo 29	29
```

**Comment by gdevenyi** [comment]
Hrm, good catch, too bad I have no idea why it misses it :-/

**Comment by pipitone** [comment]
Yes, and why it doesn't miss it locally.

One thing we can do to simplify the code and work around the string decoding issues is to default subprocess to decoding, via: 

```diff
 def command_pipe(command):
-    return Popen(shlex.split(command), stdin=PIPE, stdout=PIPE, stderr=PIPE)
+    return Popen(shlex.split(command), stdin=PIPE, stdout=PIPE, stderr=PIPE, 
+            universal_newlines=True)
```

**Comment by pipitone** [comment]
Well, I've been playing with rerunning the travis build on my own branch of this PR (https://github.com/pipitone/qbatch/tree/set-comparison-test). It seems to fail about <25% of the time and I haven't been able to replicate it locally, but when I remove `--line-buffer` from the parallel command, I don't seem to get failures...

**Comment by gdevenyi** [comment]
I wonder if it's a bug in parallel.

Perhaps we should pull down the latest version to check? It's just an untar to install.

**Comment by pipitone** [comment]
Yeah... I reproduced the travis environment: ubuntu 14.04 with parallel 20130922 and I get the same failure when using `--line-buffer`, and also when I upgrade to parallel 20140122 (same version shipped with 16.04) although it maybe happens less frequently?  Interestingly, I'm not seeing the failure using 20160822 on 14.04. 

blah.

**Comment by pipitone** [comment]
Scratch that. Just got a failure on 14.04 with 20160822.

**Comment by gdevenyi** [comment]
I'm starting to wonder if this is an interaction problem between .communicate and parallel.

It seems people sometimes have issues with Popen/communicate and missing lines...

**Comment by pipitone** [comment]
Do you have some pointers to where this problem is discussed?

**Comment by gdevenyi** [comment]
https://stackoverflow.com/questions/2432556/unbuffered-subprocess-output-last-line-missing

https://stackoverflow.com/questions/10689998/python-subprocess-losing-10-of-a-programs-stdout

**Comment by pipitone** [comment]
Hmm.. interesting. I'll take a closer look. I wonder what changed in
16.04... ;-)

How badly do you want --line-buffer?  It still seems to me that if you
want "live" output from your commands, you can just handle redirection
to a log file yourself (not friendly, but also not difficult).

**Comment by gdevenyi** [comment]
I feel pretty strongly on this.

Right now, common usage of qbatch the way its intended results in empty log files if the job runs into batch system issues. This is very user unfriendly.

**Comment by pipitone** [comment]
Fair point. I'm not sure what to do really. I'll keep investigating. 

As a stop-gap, we could use --files or --results to get parallel to dump
to files for us rather than a single shared log file. Thoughts?

**Comment by pipitone** [comment]
@gdevenyi what do you think about using `--files` or `--results` with parallel so that it dumps individual log files? I'd vote doing that, or just leaving it up to the users to write their own output redirection if they really want realtime output.

**Comment by gdevenyi** [comment]
After much testing I have determined that the problem is that GNU parallel sometimes loses lines in --line-buffer.

Tested in the latest version and it looks like it's still there. Going to attempt to report a bug.

**Pull request closed without merging**